PocketKit/Sources/SaveToPocketKit/URLExtractor.swift

```diff
@@ -0,0 +1,80 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import SharedPocketKit
+
+enum URLExtractor {
+    static func url(from itemProvider: ItemProvider) async -> String? {
+        if itemProvider.hasItemConformingToTypeIdentifier("public.url") { // We're handed a URL
+            guard let url = try? await itemProvider.loadItem(forTypeIdentifier: "public.url", options: nil) as? URL else {
+                Log.capture(message: "Unable to load URL from itemProvider")
+                return nil
+            }
+
+            return firstURL(in: url)
+        } else if itemProvider.hasItemConformingToTypeIdentifier("public.plain-text") { // We're handed a URL String
+            guard let string = try? await itemProvider.loadItem(forTypeIdentifier: "public.plain-text", options: nil) as? String else {
+                Log.capture(message: "Unable to load String from itemProvider")
+                return nil
+            }
+
+            if let firstURL = firstURL(in: string), let url = URL(string: firstURL) {
+                return self.firstURL(in: url)
+            }
```

Comment:
a bit confusing to have both the functions name `firstURL`, maybe make more descriptive?

Reply:
I personally think its expressive enough if the variable names make sense. For example:

```
let url = URL(string: "https://getpocket.com")!
var firstURL = URLExtractor.firstURL(in: url)

let urlString = "https://getpocket.com"
firstURL = URLExtractor.firstURL(in: urlString)
```

which _to me_ is more syntactically appealing than:

```
let url = URL(string: "https://getpocket.com")!
var firstURL = URLExtractor.firstURLFromURL(url)

let urlString = "https://getpocket.com"
firstURL = URLExtractor.firstURLFromString(urlString)
```

BUT…I did address your above comment and leave some comments.

Reply:
agreed, I think it was just the `firstURL` for both throwing me off, but the comments definitely helps! thanks for addressing my comments.